libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -3274,6 +3282,44 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::bytesToBytesNNConversionFunction(ArrayType const& _from, FixedBytesType const& _to)
+{
+	solAssert(_from.isByteArray(), "");
+	solAssert(_from.isDynamicallySized(), "");
+	string functionName = "convert_bytes_to_bytesNN_from_" + _from.identifier() + "_to_" + _to.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(array<?fromCalldata>, len</fromCalldata>) -> value {
+				let length := <arrayLen>(array<?fromCalldata>, len</fromCalldata>)
+				if gt(length, <fixedBytesLen>) { <panic>() }
```

Comment:
Actually interesting: So the idea is that you first have to convert this to an array slice if you want to truncate?
What if it is shorter?

Reply:
I think this should be considered carefully, since we allow `bytes4 -> bytes2` without panic.

Reply:
That was one of the questions I was asking myself as well. But it was easier to open PR and discuss here :)
Maybe we should allow only conversion if length is exactly the same?
In this case, I read `_to.numBytes` from location which can be dirty as it will read some bytes that are not necessarily zeros.
For example if we have bytes of length 31 in storage, and we convert to `bytes32` it will return actual array plus length...